### PR TITLE
Feat/validate input generator 272

### PR DESCRIPTION
**Pull request by xavier506**
Resolves [#272](https://github.com/eoscostarica/eosio-dashboard/issues/272) , Resolves [#852](https://github.com/eoscostarica/eosio-dashboard/issues/852), Resolves [#829](https://github.com/eoscostarica/eosio-dashboard/issues/859)

### CODE REVIEW

**Review by AngeloCG97 — CHANGES_REQUESTED** [review]

**Review by AngeloCG97 — COMMENTED** [review]
Please review CodeFactor alerts.

File: src/lib/ArrayTextField/index.js [hunk 1]
```diff
@@ -8,6 +8,7 @@ import TextField from '@mui/material/TextField'
 import AddIcon from '@mui/icons-material/Add'
 import Chip from '@mui/material/Chip'
 
+
```
Comment (AngeloCG97) [inline src/lib/ArrayTextField/index.js:11]: Unnecessary space

**Review by AngeloCG97 — CHANGES_REQUESTED** [review]

File: src/lib/BPJsonGenerator/EndpointsForm.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import { Validator, NODE_TYPES, NODE_EXTRA_KEYS } from '../utils'
+
+import Styles from './styles'
+
+const useStyles = makeStyles(Styles)
+
+const { hostValidation, urlInputValidation } = Validator
+
+const EndpointsForm = ({ currentNode, handleOnChange }) => {
+
```
Comment (tetogomez) [inline src/lib/BPJsonGenerator/EndpointsForm.js:16]: delete this blank space

File: src/lib/BPJsonGenerator/EndpointsForm.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import { makeStyles } from '@mui/styles'
+import TextField from '@mui/material/TextField'
+import Typography from '@mui/material/Typography'
+
+import { Validator, NODE_TYPES, NODE_EXTRA_KEYS } from '../utils'
+
+import Styles from './styles'
+
+const useStyles = makeStyles(Styles)
+
```
Comment (tetogomez) [inline src/lib/BPJsonGenerator/EndpointsForm.js:12]: delete this blank space

File: src/lib/BPJsonGenerator/EndpointsForm.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+          label="API Endpoint"
+          error={!urlInputValidation(currentNode.api_endpoint)}
+          helperText={
+            !urlInputValidation(currentNode.api_endpoint) && 'Invalid URL'
+          }
+          value={currentNode.api_endpoint || ''}
+          className={classes.formFieldForm}
+        />
+      )}
+
```
Comment (tetogomez) [inline src/lib/BPJsonGenerator/EndpointsForm.js:59]: delete this blank space

File: src/lib/BPJsonGenerator/FeaturesForm.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+import ListItemText from '@mui/material/ListItemText'
+import TextField from '@mui/material/TextField'
+import Typography from '@mui/material/Typography'
+
+import { NODE_EXTRA_KEYS } from '../utils'
+
+import Styles from './styles'
+
+const useStyles = makeStyles(Styles)
+
```
Comment (tetogomez) [inline src/lib/BPJsonGenerator/FeaturesForm.js:16]: delete this blank space

File: src/lib/utils/schemas.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  return requiredValidation(value) && urlInputValidation(value)
+}
+
+const locationSchema = {
+  name: { isValid: requiredValidation, isRequired: true, message: 'The name is required' },
+  country: { isValid: countryValidation, isRequired: true, message: 'The country code must be two letters' },
+  latitude: { isValid: latitudeValidation, isRequired: false, message: 'The latitude range is between -90 and 90' },
+  longitude: { isValid: longitudeValidation, isRequired: false, message: 'The longitude range is between -180 and 180' }
+}
+
```
Comment (tetogomez) [inline src/lib/utils/schemas.js:26]: delete this blank space

File: src/lib/utils/schemas.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+const nodeSchema = {
+  node_type: { isValid: nodeTypeValidation, isRequired: true },
+  full: { isValid: isBoolean, isRequired: true },
+  location: { isValid: (obj) => validate(obj, locationSchema), isRequired: false },
+  api_endpoint: { isValid: urlInputValidation, isRequired: false },
+  ssl_endpoint: { isValid: urlInputValidation, isRequired: false },
+  p2p_endpoint: { isValid: hostValidation, isRequired: false }
+}
+
```
Comment (tetogomez) [inline src/lib/utils/schemas.js:35]: delete this blank space

File: src/lib/utils/schemas.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  candidate_name: { isValid: requiredValidation, isRequired: true, message: 'Candidate Name is required' },
+  website: { isValid: requiredUrlValidation, isRequired: true, message: 'Invalid URL' },
+  email: { isValid: emailInputValidation, isRequired: true, message: 'Invalid Email format' },
+  code_of_conduct: { isValid: requiredUrlValidation, isRequired: true, message: 'Invalid URL' },
+  ownership_disclosure: { isValid: requiredUrlValidation, isRequired: true, message: 'Invalid URL' },
+  chain_resources: { isValid: urlInputValidation, isRequired: false },
+  other_resources: { isValid: urlInputValidation, isRequired: false },
+  location: { isValid: (obj) => validate(obj, locationSchema), isRequired: true },
+}
+
```
Comment (tetogomez) [inline src/lib/utils/schemas.js:46]: delete this blank space

**Review by tetogomez — CHANGES_REQUESTED** [review]
See the comments

**Review by AngeloCG97 — APPROVED** [review]
LGTM

File: src/lib/BPJsonGenerator/BoxDropzone.js [hunk 9]
```diff
@@ -20,18 +24,39 @@ const Dropzone = ({ onSubmit }) => {
...
+        } catch (error) {
+          if (error instanceof SyntaxError) {
+            setErrorMessage("The file does not have the correct JSON format")
+          } else {
+            setErrorMessage(error.message)
+          }
+          setOpenModal(true)
+        }
+      }
+
```
Comment (leisterfrancisco) [inline src/lib/BPJsonGenerator/BoxDropzone.js:41]: remove empty line

File: src/lib/BPJsonGenerator/BoxDropzone.js [hunk 10]
```diff
@@ -20,18 +24,39 @@ const Dropzone = ({ onSubmit }) => {
...
+      if (acceptedFiles[0] !== lastFile) {
+        try {
+          onSubmit(JSON.parse(e.target.result))
+          setLastFile(acceptedFiles[0])
+        } catch (error) {
+          if (error instanceof SyntaxError) {
+            setErrorMessage("The file does not have the correct JSON format")
+          } else {
+            setErrorMessage(error.message)
+          }
```
Comment (leisterfrancisco) [inline src/lib/BPJsonGenerator/BoxDropzone.js:37]: Recommendation

```
setErrorMessage(error instanceof SyntaxError ? "The file does not have the correct JSON format" : error.message)
```

**Review by leisterfrancisco — CHANGES_REQUESTED** [review]

**Review by fvives99 — APPROVED** [review]
LGTM

File: src/lib/ArrayTextField/index.js [hunk 11]
```diff
@@ -7,7 +7,6 @@ import InputAdornment from '@mui/material/InputAdornment'
 import TextField from '@mui/material/TextField'
 import AddIcon from '@mui/icons-material/Add'
 import Chip from '@mui/material/Chip'
-
```
Comment (tetogomez) [inline src/lib/ArrayTextField/index.js:10]: This blank space is necessary.
clean code rule 

`order import by a group`

first group: third-party libraries
second group: project libraries (components, config, utils, etc)
third group: files on the same level (styles files or a specific component used only here)

so, these groups should be separated by a blank space

File: src/lib/ArrayTextField/index.js [hunk 12]
```diff
@@ -17,14 +16,16 @@ const ArrayTextField = ({
...
+  ArrayValidator,
   ...props
 }) => {
   const classes = useStyles()
   const [items, setItems] = useState(value)
   const [item, setItem] = useState('')
 
   const handleOnAddItem = () => {
-    if (!item) {
+
```
Comment (tetogomez) [inline src/lib/ArrayTextField/index.js:27]: delete blank space

File: src/lib/BPJsonGenerator/EndpointsForm.js [hunk 13]
```diff
@@ -0,0 +1,72 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import { makeStyles } from '@mui/styles'
+import TextField from '@mui/material/TextField'
+import Typography from '@mui/material/Typography'
+import { Validator, NODE_TYPES, NODE_EXTRA_KEYS } from '../utils'
+import Styles from './styles'
```
Comment (tetogomez) [inline src/lib/BPJsonGenerator/EndpointsForm.js:7]: same here, please check the comment about `order import by a group`

File: src/lib/BPJsonGenerator/EndpointsForm.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+const useStyles = makeStyles(Styles)
+const { hostValidation, urlInputValidation } = Validator
+const EndpointsForm = ({ currentNode, handleOnChange }) => {
+  const classes = useStyles()
+  if (currentNode.node_type === NODE_TYPES.PRODUCER || currentNode.node_type === '') {
+    return <></>
+  }
```
Comment (tetogomez) [inline src/lib/BPJsonGenerator/EndpointsForm.js:15]: after if declaration always add blank space

File: src/lib/BPJsonGenerator/FeaturesForm.js [hunk 15]
```diff
@@ -0,0 +1,123 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+import { makeStyles } from '@mui/styles'
+import Checkbox from '@mui/material/Checkbox'
+import Chip from '@mui/material/Chip'
+import MenuItem from '@mui/material/MenuItem'
+import ListItemText from '@mui/material/ListItemText'
+import TextField from '@mui/material/TextField'
+import Typography from '@mui/material/Typography'
+import { NODE_EXTRA_KEYS } from '../utils'
```
Comment (tetogomez) [inline src/lib/BPJsonGenerator/FeaturesForm.js:10]: `order import by a group`

**Review by tetogomez — CHANGES_REQUESTED** [review]

**Review by tetogomez — APPROVED** [review]
LGTM

good job!

**Pull request merged**